src/Synthesiser.cpp

```diff
@@ -829,6 +829,216 @@ void Synthesiser::emitCode(std::ostream& out, const RamStatement& stmt) {
+                    break;
+                }
+
+                case AggregateOp::MEAN:
+                case AggregateOp::FSUM:
+                case AggregateOp::USUM:
+                case AggregateOp::COUNT:
+
+                case AggregateOp::SUM: {
+                    op = "+";
+                    break;
+                }
+                default: fatal("Unhandled aggregate operation");
+            }
+            // res0 stores the aggregate result
+            std::string sharedVariable = "res0";
+
+            std::string type;
+            switch (getTypeAttributeAggregate(aggregate.getFunction())) {
+                case TypeAttribute::Signed: type = "RamSigned"; break;
+                case TypeAttribute::Unsigned: type = "RamUnsigned"; break;
+                case TypeAttribute::Float: type = "RamFloat"; break;
+
+                case TypeAttribute::Symbol:
+                case TypeAttribute::Record: type = "RamDomain"; break;
+            }
+            out << type << " res0 = " << init << ";\n";
+            if (aggregate.getFunction() == AggregateOp::MEAN) {
+                out << type << " res1 = 0;\n";
+                sharedVariable += ", res1";
+            }
+
+            // check whether there is an index to use
+            if (keys.empty()) {
+                out << "#pragma omp parallel for reduction(" << op << ":" << sharedVariable << ")\n";
+                out << "for(const auto& env" << identifier << " : "
+                    << "*" << relName << ") {\n";
+            } else {
+                const auto& patternsLower = aggregate.getRangePattern().first;
+                const auto& patternsUpper = aggregate.getRangePattern().second;
+
+                out << preamble.str();
+
+                out << "const " << tuple_type << " lower{{";
+                out << join(patternsLower.begin(), patternsLower.begin() + arity, ",", recWithDefault);
+                out << "}};\n";
+
+                out << "const " << tuple_type << " upper{{";
+                out << join(patternsUpper.begin(), patternsUpper.begin() + arity, ",", recWithDefault);
+                out << "}};\n";
+
+                out << "auto range = " << relName << "->"
+                    << "lowerUpperRange_" << keys << "(lower,upper," << ctxName << ");\n";
+
+                out << "auto part = range.partition();\n";
+                out << "#pragma omp parallel for reduction(" << op << ":" << sharedVariable << ")\n";
+                // iterate over each part
+                out << "for (auto it = part.begin(); it < part.end(); ++it) {\n";
+                // iterate over tuples in each part
+                out << "for (const auto& env" << identifier << ": *it) {\n";
+            }
+
+            // produce condition inside the loop if necessary
+            if (!isRamTrue(&aggregate.getCondition())) {
+                out << "if( ";
+                visit(aggregate.getCondition(), out);
+                out << ") {\n";
+            }
```

Comment:
This could be simplified to
```suggestion
            out << "if( ";
            visit(aggregate.getCondition(), out);
            out << ") {\n";
```

The output may have a pointless `if (true)`, but the compiler will take care of that and it makes the code here simpler.